File: decoder.go
package dotbeam

import "errors"

var (
	ErrIncompleteData = errors.New("dotbeam: incomplete data, not all frames received")
	ErrInvalidFrame   = errors.New("dotbeam: invalid frame header")
)

// Decoder reassembles data from captured dotbeam frames.
type Decoder struct {
	config   Config
	frames   map[int][]byte // frame index → payload
	total    int
	received int
}

// NewDecoder creates a new decoder with the given config.
func NewDecoder(config Config) *Decoder {
	return &Decoder{
		config: config,
		frames: make(map[int][]byte),
	}
}

// AddFrame processes a decoded frame's dot values and stores the payload.
// Returns true if all frames have been received.
func (d *Decoder) AddFrame(dots []Dot) (bool, error) {
	if len(dots) == 0 {
		return false, ErrInvalidFrame
	}

	// Convert dot values back to bytes
	data := d.dotsToBytes(dots)
	if len(data) < 2 {
		return false, ErrInvalidFrame
	}

	frameIndex := int(data[0])
	frameTotal := int(data[1])

	if frameTotal == 0 {
		return false, ErrInvalidFrame
	}

	d.total = frameTotal
	payload := data[2:]

	if _, exists := d.frames[frameIndex]; !exists {
		d.frames[frameIndex] = payload
		d.received++
	}

	return d.received >= d.total, nil
}

// Data returns the reassembled data. Returns error if incomplete.
func (d *Decoder) Data() ([]byte, error) {
	if d.received < d.total {
		return nil, ErrIncompleteData
	}

	var result []byte
	for i := 0; i < d.total; i++ {
		payload, ok := d.frames[i]
		if !ok {
			return nil, ErrIncompleteData
		}
		result = append(result, payload...)
	}

	return result, nil
}

// Progress returns the fraction of frames received (0.0 to 1.0).
func (d *Decoder) Progress() float64 {
	if d.total == 0 {
		return 0
	}
	return float64(d.received) / float64(d.total)
}

// Reset clears all received frames.
func (d *Decoder) Reset() {
	d.frames = make(map[int][]byte)
	d.total = 0
	d.received = 0
}

// dotsToBytes converts dot values back into a byte slice.
func (d *Decoder) dotsToBytes(dots []Dot) []byte {
	bitsPerDot := d.config.BitsPerDot
	bits := make([]uint8, len(dots)*bitsPerDot)

	for i, dot := range dots {
		for b := bitsPerDot - 1; b >= 0; b-- {
			bits[i*bitsPerDot+(bitsPerDot-1-b)] = (dot.Value >> b) & 1
		}
	}

	// Pack bits into bytes
	numBytes := len(bits) / 8
	data := make([]byte, numBytes)
	for i := 0; i < numBytes; i++ {
		var val byte
		for j := 0; j < 8; j++ {
			val = (val << 1) | bits[i*8+j]
		}
		data[i] = val
	}

	return data
}

File: js/src/decoder.js
/**
 * Decoder — mirrors Go decoder.go
 */

import { DEFAULT_CONFIG } from "./index.js";

export class Decoder {
  constructor(config = DEFAULT_CONFIG) {
    this.config = config;
    this.frames = new Map();
    this.total = 0;
    this.received = 0;
  }

  /**
   * Add a frame's dots. Returns true if all frames received.
   * @param {Array<{value:number}>} dots
   * @returns {boolean}
   */
  addFrame(dots) {
    if (!dots || dots.length === 0) {
      throw new Error("dotbeam: invalid frame");
    }

    const data = this._dotsToBytes(dots);
    if (data.length < 2) {
      throw new Error("dotbeam: invalid frame header");
    }

    const frameIndex = data[0];
    const frameTotal = data[1];
    if (frameTotal === 0) {
      throw new Error("dotbeam: invalid frame total");
    }

    this.total = frameTotal;

    if (!this.frames.has(frameIndex)) {
      this.frames.set(frameIndex, data.slice(2));
      this.received++;
    }

    return this.received >= this.total;
  }

  /**
   * Get reassembled data. Throws if incomplete.
   * @returns {Uint8Array}
   */
  data() {
    if (this.received < this.total) {
      throw new Error("dotbeam: incomplete data");
    }

    const chunks = [];
    let totalLength = 0;
    for (let i = 0; i < this.total; i++) {
      const payload = this.frames.get(i);
      if (!payload) throw new Error("dotbeam: incomplete data");
      chunks.push(payload);
      totalLength += payload.length;
    }

    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }
    return result;
  }

  progress() {
    if (this.total === 0) return 0;
    return this.received / this.total;
  }

  reset() {
    this.frames.clear();
    this.total = 0;
    this.received = 0;
  }

  _dotsToBytes(dots) {
    const bpd = this.config.bitsPerDot;
    const bits = [];
    for (const dot of dots) {
      for (let b = bpd - 1; b >= 0; b--) {
        bits.push((dot.value >> b) & 1);
      }
    }

    const numBytes = Math.floor(bits.length / 8);
    const data = new Uint8Array(numBytes);
    for (let i = 0; i < numBytes; i++) {
      let val = 0;
      for (let j = 0; j < 8; j++) {
        val = (val << 1) | bits[i * 8 + j];
      }
      data[i] = val;
    }
    return data;
  }
}
